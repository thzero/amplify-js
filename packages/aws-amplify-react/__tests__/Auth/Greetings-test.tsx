import Auth from '@aws-amplify/auth';
import { Hub } from '@aws-amplify/core';
import Greetings from '../../src/Auth/Greetings';
import * as React from 'react';
import AmplifyTheme from '../../src/AmplifyTheme';
import AuthPiece from '../../src/Auth/AuthPiece';
import {
	Header,
	Footer,
	Input,
	Button,
} from '../../src/Amplify-UI/Amplify-UI-Components-React';

const acceptedStates = ['signedIn'];

const deniedStates = [
	'signIn',
	'signedUp',
	'signedOut',
	'forgotPassword',
	'signUp',
	'confirmSignIn',
	'confirmSignUp',
	'verifyContact',
];

describe('Greetings', () => {
	describe('normal case', () => {
		test('render correctly with authState signedIn', () => {
			const wrapper = shallow(<Greetings />);
			for (let i = 0; i < acceptedStates.length; i += 1) {
				wrapper.setProps({
					authState: acceptedStates[i],
					authData: {
						attributes: {
							name: 'username',
						},
					},
					theme: 'theme',
				});
				expect(wrapper).toMatchSnapshot();
			}
		});

		test('render correctly with hide', () => {
			const wrapper = shallow(<Greetings />);
			for (let i = 0; i < acceptedStates.length; i += 1) {
				wrapper.setProps({
					authState: acceptedStates[i],
					theme: 'theme',
					hide: [Greetings],
				});
				expect(wrapper).toMatchSnapshot();
			}
		});
	});

	test('render corrently with other authStates', () => {
		const wrapper = shallow(<Greetings />);

		for (let i = 0; i < deniedStates.length; i += 1) {
			wrapper.setProps({
				authState: deniedStates[i],
				theme: 'theme',
			});

			expect(wrapper).toMatchSnapshot();
		}
	});

	describe('Greetings lifecycle', () => {
		test('componentDidMount', async () => {
			const spy = jest.spyOn(Greetings.prototype, 'componentDidMount');
			const greetings = mount(<Greetings />).instance();
			greetings.componentDidMount();
			expect(greetings._isMounted).toBeTruthy();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('componentWillUnmount', async () => {
			const wrapper = shallow(<Greetings />);
			const greetings = wrapper.instance();
			const componentWillUnmount = jest.spyOn(
				greetings,
				'componentWillUnmount'
			);
			wrapper.unmount();
			expect(greetings._isMounted).toBeFalsy();
			expect(componentWillUnmount).toHaveBeenCalled();
		});
	});

	describe('findState tests', () => {
		test('findState is called', () => {
			const spy = jest.spyOn(Greetings.prototype, 'findState');
			const greetings = mount(<Greetings />).instance();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser is not called if auth props are present', () => {
			const spy = jest
				.spyOn(Auth, 'currentAuthenticatedUser')
				.mockImplementationOnce(() => {
					return new Promise((res, rej) => {
						res({
							user: {},
						});
					});
				});
			const props = {
				authState: 'signedIn',
				authData: {},
			};
			const wrapper = shallow(<Greetings {...props} />);
			const greetings = wrapper.instance();
			expect(spy).not.toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser is called if auth props are not present', () => {
			const spy = jest
				.spyOn(Auth, 'currentAuthenticatedUser')
				.mockImplementationOnce(() => {
					return new Promise((res, rej) => {
						res({
							user: {},
						});
					});
				});
			const props = {};
			const wrapper = shallow(<Greetings {...props} />);
			const greetings = wrapper.instance();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser results in state being set', async () => {
			const spy = jest.spyOn(Greetings.prototype, 'setState');
			const props = {};
			const wrapper = shallow(<Greetings {...props} />);
			const greetings = wrapper.instance();
			await greetings.findState();
			expect(greetings.state.stateFromStorage).toEqual(true);
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
	});

	describe('onHubCapsule tests', () => {
		test('onHubCapsule is present', () => {
			const greetings = mount(<Greetings />).instance();
			expect(greetings.onHubCapsule).toBeTruthy();
		});
		test('onHubCapsule is called on a Hub event', () => {
			const spy = jest.spyOn(Greetings.prototype, 'onHubCapsule');
			const greetings = mount(<Greetings />).instance();
			Hub.dispatch('auth', { event: 'test' });
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('onHubCapsule should setState with authState = "signedIn" when "signIn" auth event fires', () => {
			const spy = jest.spyOn(Greetings.prototype, 'setState');
			const greetings = mount(<Greetings />).instance();
			Hub.dispatch('auth', { event: 'signIn', data: { foo: 'bar' } });
			expect(spy).toHaveBeenCalledWith({
				authState: 'signedIn',
				authData: { foo: 'bar' },
			});
			spy.mockClear();
		});
		test('onHubCapsule should setState with authState = "signIn" when "customSignOut" auth event fires', () => {
			const spy = jest.spyOn(Greetings.prototype, 'setState');
			const greetings = mount(<Greetings />).instance();
			Hub.dispatch('auth', { event: 'signOut' });
			expect(spy).toHaveBeenCalledWith({
				authState: 'signIn',
			});
			spy.mockClear();
		});
	});
});
