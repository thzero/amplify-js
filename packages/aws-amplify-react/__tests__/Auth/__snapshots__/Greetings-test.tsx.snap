// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Greetings normal case render correctly with authState signedIn 1`] = `
<Component
  data-test="greetings-nav-bar"
  theme="theme"
>
  <Component
    data-test="greetings-nav"
    theme="theme"
  >
    <Component
      data-test="greetings-nav-right"
      theme="theme"
    >
      <span>
        <Component
          theme="theme"
        >
          Hello username
        </Component>
        <SignOut
          authData={
            Object {
              "attributes": Object {
                "name": "username",
              },
            }
          }
          authState="signedIn"
          theme="theme"
        />
      </span>
    </Component>
  </Component>
</Component>
`;

exports[`Greetings normal case render correctly with hide 1`] = `""`;

exports[`Greetings render corrently with other authStates 1`] = `""`;

exports[`Greetings render corrently with other authStates 2`] = `""`;

exports[`Greetings render corrently with other authStates 3`] = `""`;

exports[`Greetings render corrently with other authStates 4`] = `""`;

exports[`Greetings render corrently with other authStates 5`] = `""`;

exports[`Greetings render corrently with other authStates 6`] = `""`;

exports[`Greetings render corrently with other authStates 7`] = `""`;

exports[`Greetings render corrently with other authStates 8`] = `""`;
